support range int

sploving opened the pull request:
This commit support range int. For instance,   a filed x with int8 [3, 6], means that x could be 3, 4, 5. This is useful for some system calls ,which only accepts a range integer.

dvyukov commented:
Hi,

Thanks for working on this.
Can you provide some examples of syscalls that need this?
Do you plan to use this ranged int in the description?

dvyukov commented on prog/rand.go:
```diff
@@ -737,6 +744,9 @@ func (r *randGen) generateArg(s *state, typ sys.Type, dir ArgDir, sizes map[stri
 		return dataArg([]byte(a.Val)), constArg(uintptr(len(a.Val))), nil
 	case sys.IntType:
 		v := r.randInt()
+		if a.IsRange == true {
```
There is already IntKind that controls generation.
Please add IntRange kind instead of IsRange bool, and then handle it in the switch below.

dvyukov commented on sys/decl.go:
```diff
@@ -391,6 +391,8 @@ type IntType struct {
 	TypeSize uintptr
 	Kind     IntKind
+	IsRange  bool
+	IntRange []uintptr
```
Using a slice here is confusing. It does not say how many entries must be in the slice, while we always expect a fixed number of elements. It can also have wrong number of elements, so there is also a temptation to check that a slice has correct number of arguments, but it will only unnecessary bloat code.
Also range can have negative begin/end.
Just use:
```
int64 rangeBegin
int64 rangeEnd
```

sploving replied:
randInt and rand64 both return uintptr in prog/rand.go . so rangeBegin and rangeEnd should also be uintptr?  since the range may be for  int8 ,  int16, int32, not only for int64. Does the negative also be presented by  uintptr?  Could you tell me the difference between uintptr and int64 ? Thanks

dvyukov replied:
randInt and rand64 return uintptr, but representation of that uintptr can mean negative numbers if treated as a signed integer (e.g. 0xffffffffffff means -1).

uintptr and int64 are pretty much the same as in C. uintptr is an unsigned integer with size equal to size of a pointer. int64 is a signed 64-bit integer.

OK, I think the simplest thing for now is to parse the numbers as signed, but then convert them to uintptr:
```
    var lo, hi int64
    if _, err := fmt.Sscanf(s, "%d-%d", &lo, &hi); err != nil {
        failf("failed to parse int range: %v (%v)", s, err)
    }
    fmt.Fprintf(out, "IntType{%v, Kind: IntRange, TypeSize: %v, rangeBegin: %v, rangeEnd: %v}", common(), typeToSize(typ), uintptr(lo), uintptr(hi))
```

dvyukov commented on prog/rand.go:
```diff
@@ -71,6 +71,13 @@ func (r *randGen) randInt() uintptr {
+	v := r.randInt()
+	intRange := end - begin
+	offset := v % intRange
+	return begin + offset
```
This all can be replaced with a one liner:
```
return begin + r.rand(end - begin)
```

sploving replied:
How to generate inclusive range?  begin + r.rand(end - begin) does not include end.

dvyukov replied:
begin + r.rand(end - begin + 1)?

dvyukov commented on sysgen/sysgen.go:
```diff
@@ -288,10 +288,13 @@ func generateArg(name, typ string, a []string, structs map[string]Struct, unname
 	case "int8", "int16", "int32", "int64", "intptr":
-		if want := 0; len(a) != want {
-			failf("wrong number of arguments for %v arg %v, want %v, got %v", typ, name, want, len(a))
+		if len(a) == 2 {
```
It is meant to be inclusive range, or exclusive range?
I think it's more intuitive to have an inclusive range.
Please extend description of int type on top of sys/sys.txt.

I agree that using flags for ranges is tedious.
But my concern is that once we give some meaning to arguments of a type, it will be difficult to change it (some users can have private syscall descriptions). E.g. if later we come up with better meaning for int arguments, we won't be able to change it.
I think that ultimately we want to support not only ranges, but arbitrary sets of values. For example:
```
foo int32[0, 5-10, 50, 100-200]
```
So please change the range format to `[begin-end]`. We will be able to extend such form the more general form.
You can use the following code for parsing:
https://play.golang.org/p/CQwd_G6keY

sploving replied:
Yes. the form foo int32[0, 5-10, 50, 100-200] is what we want. It is useful for users. I cannot open the link https://play.golang.org/p/CQwd_G6keY   It is forbidden.  Could you help me? Thanks.

dvyukov replied:
> Yes. the form foo int32[0, 5-10, 50, 100-200] is what we want.

You don't need to support the full form right now (if you need just the range now). If we change the range format to "[begin-end]", we can always extend this to the full form later.

> I cannot open the 

Sorry. Sure. The code is:
```go
package main

import (
	"fmt"
)

func main() {
	s := "10-15"
	var lo, hi int64
	if _, err := fmt.Sscanf(s, "%d-%d", &lo, &hi); err != nil {
		failf("failed to parse int range: %v (%v)", s, err)
	}
	if lo >= hi {
		failf("bad int range: %v", s)
	}
	fmt.Println(lo, hi)
}
```

dvyukov commented on sysgen/sysgen.go:
```diff
@@ -288,10 +288,13 @@ func generateArg(name, typ string, a []string, structs map[string]Struct, unname
-		if want := 0; len(a) != want {
-			failf("wrong number of arguments for %v arg %v, want %v, got %v", typ, name, want, len(a))
+		if len(a) == 2 {
+			range_begin := a[0]
```
Go coding style is to use camelCase instead of c++_style. So:
s/range_begin/rangeBegin/

sploving commented:
2016-08-09 16:21 GMT+08:00 Dmitry Vyukov <notifications@github.com>:

> Hi,
>
> Thanks for working on this.
> Can you provide some examples of syscalls that need this?
>
Here are some examples:
1. ioctl$EVIOCSKEYCODE_V2
http://lxr.free-electrons.com/source/drivers/input/evdev.c#L895
the range for ke.len is [0, 32].
2. ioctl$SNDRV_TIMER_IOCTL_TREAD
http://lxr.free-electrons.com/source/sound/core/timer.c#L1863
  xarg could be 0 or not 0. So [0, 1] (inclusive range) is enough for xarg.
3. ioctl$RNDADDENTROPY
http://lxr.free-electrons.com/source/drivers/char/random.c#L1553
ent_count should not smaller than 0. So the range for ent_count is  [0,
0x7fffffff]

Use the ranged int, we could reduce fuzzy time and get more code coverage
efficiently.
Also we could avoid crash if we know some integer input crash the kernel.
We just exclude the integer that causes the kernel panic and use the other
ranged integer to continue testing the kernel without any need to
re-compile the kernel.

> Do you plan to use this ranged int in the description?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/syzkaller/pull/53#issuecomment-238485948>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAsDDSwrnpD4dTWHcCb-3MMprkWFhGCXks5qeDh_gaJpZM4JfDRa>
> .
>



-- 
     Best Regards,
     Baozeng Ding

dvyukov commented:
> Also we could avoid crash if we know some integer input crash the kernel.
We just exclude the integer that causes the kernel panic and use the other
ranged integer to continue testing the kernel without any need to
re-compile the kernel.


I think we need to generate integers outside of the range as well, but with lower probability. This provide both efficiency and completeness.
Re avoiding crashing kernel: I don't think altering syscall description is a productive way to avoid crashes. Crashes can be caused by lots of other things, we can't avoid all them with ranged ints. And we already have 2 mechanisms to temporary work around known kernel bugs: suppressions and disabled syscalls in config. These mechanisms are more general and does not require patching syscall description, regenerating and rebuilding everything. But eventually all bugs need to be fixed.
Why don't these mechanisms work for you.

dvyukov commented on prog/rand.go:
```diff
@@ -71,6 +71,13 @@ func (r *randGen) randInt() uintptr {
 }
 
+func (r *randGen) randRangeInt(begin uintptr, end uintptr) uintptr {
+	v := r.randInt()
```
Add:
```
if r.oneOf(100) {
  return r.randInt()
}
```
here. It will provide more coverage while ensuring efficiency.

sploving commented:
2016-08-09 19:37 GMT+08:00 Dmitry Vyukov <notifications@github.com>:

> Also we could avoid crash if we know some integer input crash the kernel.
> We just exclude the integer that causes the kernel panic and use the other
> ranged integer to continue testing the kernel without any need to
> re-compile the kernel.
>
> I think we need to generate integers outside of the range as well, but
> with lower probability. This provide both efficiency and completeness.
> Re avoiding crashing kernel: I don't think altering syscall description is
> a productive way to avoid crashes. Crashes can be caused by lots of other
> things, we can't avoid all them with ranged ints. And we already have 2
> mechanisms to temporary work around known kernel bugs: suppressions and
> disabled syscalls in config. These mechanisms are more general and does not
> require patching syscall description, regenerating and rebuilding
> everything. But eventually all bugs need to be fixed.
> Why don't these mechanisms work for you.
>
With qemu/Linux type, it is okay, since the latest kernel supports KASAN
well.  But for adb/android type, I have not ported KASAN very well, so the
suppressions mechanism does not work well currently. And disabled syscalls
mechanism is too coarse.  Altering syscall description is a complementary
approach.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/syzkaller/pull/53#issuecomment-238527450>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAsDDeuQXEejnZsQJ1iDRXdfiySa4Vgwks5qeGZjgaJpZM4JfDRa>
> .
>



-- 
     Best Regards,
     Baozeng Ding

dvyukov commented:
Then you need to finish porting KASAN. You won't get too far until it works well anyway. We should not work around poorly working KASAN in a particular older kernel tree with syzkaller hacks.
If you want to locally temporary disable some particular values, you can use const argument type. Const is meant to be exactly that single value, without variations.
For the rest of the argument types we should move in the opposite direction -- generate even more variations.

sploving commented:
I updated the code. I use '~' to mean the range, and '-' to mean a negative number. So that we could use foo int32[-10~10].

dvyukov commented:
I've merged this, but I fixed several things afterwards:
https://github.com/google/syzkaller/commits/master
In particular:
1. Generate values outside of the range once in a while. You can comment out that part locally if you need strict ranges. But generally fuzzer should not be restricted in values it generates.
2. I've changed range syntax to use ':' instead of '~'. Colon in used in several well known range notations (e.g. Mathlab, Excel, Go).  While I never seen '~' to denote ranges.

sploving commented:
Thanks.

2016-08-14 0:23 GMT+08:00 Dmitry Vyukov <notifications@github.com>:

> I've merged this, but I fixed several things afterwards:
> https://github.com/google/syzkaller/commits/master
> In particular:
> 1. Generate values outside of the range once in a while. You can comment
> out that part locally if you need strict ranges. But generally fuzzer
> should not be restricted in values it generates.
> 2. I've changed range syntax to use ':' instead of '~'. Colon in used in
> several well known range notations (e.g. Mathlab, Excel, Go). While I never
> seen '~' to denote ranges.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/syzkaller/pull/53#issuecomment-239629125>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAsDDZMFKyF8DAbGwf1sgMT40T23f4Vpks5qfe98gaJpZM4JfDRa>
> .
>



-- 
     Best Regards,
     Baozeng Ding
